interfaces/apparmor/spec.go

```diff
@@ -173,6 +173,96 @@ func isProbablyPresent(path string) bool {
 	return path == "/" || path == "/snap" || path == "/var" || path == "/var/snap" || path == "/tmp" || path == "/usr" || path == "/etc"
 }
 
+// chopTree takes a path and depth and returns two apparmor path expressions.
+//
+// The returned expressions are referred to as left and right.
+//
+// The left expression describes directories at depth up to assumedPrefixDepth
+// and can be used to grant read permission to them (by appending the string
+// "r, "). This corresponds to an assumption about those directories are
+// present in the system and are just traversed.
+//
+// The right expression describes the remaining directories and the leaf entry
+// (either file or directory) and is somewhat subtle. At each depth level the
+// expression describes all the files and directories at that level. Coupled
+// with the string "rw ," it can be used to create a rule that allows write
+// access to any file therein, but not deeper.
+//
+// For example, with path: "/foo/bar/froz/baz" and depth 2 the result is:
+// "/{,foo/{,bar/}}" and "/foo/bar/{*,*/,froz/{*,*/,baz}}". Coupled with the
+// aforementioned constants this translates to the following apparmor rules:
+//
+//   /{,foo/{,bar/}} r,
+//   /foo/bar/{*,*/,froz/{*,*/,baz}} rw,
+//
+// Those rules allow the code to open (to traverse securely) /, then foo, then
+// bar and then create _either_ 1) any file in /foo/bar/ or 2) any directory
+// in /foo/bar/ or 3) the file froz/ in /foo/bar/. Subsequently, but just
+// inside the directory /foo/bar/froz/ we can create any file, any directory
+// or the file baz. Technically the last rule has some redundancy as baz is
+// covered by * but this is retained for consistence.
+//
+// Those rules are useful for constructing the apparmor profile for a writable
+// mimic that needs to be present in a specific directory (e.g. in
+// /foo/bar/froz) assuming that part of that directory already exists (e.g.
+// /foo/) but may need to be created earlier (e.g. in /foo/bar/) because
+// /foo/bar/ is a read-only file system.
+//
+// The mimic works by mounting a tmpfs over the mimicked directory and then
+// re-creating empty files and directories as mount points for the subsequent
+// bind- mount operations to latch onto. This is the reason for the {*,*/,..}
+// part of the expression (we need it to make the empty placeholders).
+func chopTree(path string, assumedPrefixDepth int) (string, string) {
```

Comment:
Non iterative alternative https://play.golang.org/p/nsqzu1JuxWB

Reply:
Does it pass tests? :-)

Reply:
Yeah, I used these test cases https://github.com/snapcore/snapd/pull/5081/files#diff-16cc40a4b635e1e0e4c9d520342fd500R291 Just to be clear, not a blocker or anything, it was just fun to write.